src/image.rs

```diff
@@ -0,0 +1,1192 @@
+//! Utility code for working with images.
+//!
+//! This module contains the [`Image`] struct which represents an arbitrary image in
+//! `ImageFormat::ZPixmap`. If you do not know what `ZPixmap` means, then rest assured that you do
+//! not want to know the details. It suffices to know that the values of the individual pixels are
+//! saved one after another in memory.
+//!
+//! An [`Image`] can be converted to a different internal representation. [`Image::native`]
+//! converts it to the native format of the X11 server. These conversions do not change the actual
+//! content of the image, but only the way that it is laid out in memory (e.g. byte order and
+//! padding). Specifically, there is no support for converting an image to another `depth`.
+
+// For future readers:
+//
+// - Z-Pixmap means that the pixels are right next to each other. I.e. first you have the value of
+//   the pixel at (0, 0), then (1, 0), .... At the end of the row, there might be some padding
+//   before the next row begins.
+// - XY-Bitmap consists of individual bits. The bits are assigned colors via a GC's foreground and
+//   background color when uploading to the X11 server.
+// - XY-Pixmap consists of bit planes. This means you first get the first bit of each pixel,
+//   stuffed together into bytes. After all of the first pixels are represented, the second plane
+//   begins with the second bit of each pixel etc.
+
+use std::borrow::Cow;
+use std::convert::{TryFrom, TryInto};
+
+use crate::connection::Connection;
+use crate::cookie::VoidCookie;
+use crate::errors::{ConnectionError, ParseError, ReplyError};
+use crate::protocol::xproto::{
+    Drawable, Format, Gcontext, GetImageReply, GetImageRequest, ImageFormat, ImageOrder,
+    PutImageRequest, Setup, VisualClass, Visualtype,
+};
+
+/// The description of a single color component.
+///
+/// For example, in an RGB image, pixels are often saved as `0xRRGGBB`, where each letter
+/// represents the respective color component. In the example, green has a `width` of 8 (it takes
+/// up 8 bits) and a `shift` of `16` (there are 16 less significant bits beyond it). This info is
+/// represented as a `ColorComponent`.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub struct ColorComponent {
+    /// Number of bits for the component
+    width: u8,
+    /// Offset in an u32 of the component
+    shift: u8,
```

Comment:
These fields are private, so it would be nice to have functions to get their value.